### DISCUSSION TITLE
关于使用 LaTeX3 的重构工作

### DISCUSSION THREAD

**Discussion by atxy-blip** [tuna/thuthesis#827] [Ideas]
`thuthesis` 发布至今有近 20 年，其中积累了诸位前辈的巨量贡献，形成了如今的规模。当下，[LaTeX3 编程接口](https://ctan.org/pkg/l3kernel)业已基本成熟，小弟斗胆建议使用 L3 对模板进行重构，以适应新的发展趋势。在这里希望与大家探讨一下可行性和方法问题。

### 可行性问题

使用 L3 编写的宏包和文档类已有不少，`ctex` 宏集本身就可以作为典型例子。我做过的一些测试表明，`thuthesis` 本身已足够高效，L3 并不会带来足够可观的编译速度提升。采用新的代码风格和接口更多是出于可维护性的考虑：L3 引入的变量类型、展开类型符号以及库函数使得命令的执行过程更加清晰。

### 方向问题

我私下征求了当前维护者 @zepinglee 的意见，他认为应该新开分支逐步进行重构，以利用当前的 `l3build` 构建测试，同时维护一份单一的 `cls` 文件降低用户的升级难度。我个人更倾向于新开仓库另起炉灶，这样可以利用 DosStrip 的特性实现模块化的配置文件，个人名下的仓库（[`atxy-blip/thuthesis3`](https://github.com/atxy-blip/thuthesis3)）尝试对此进行了展示。

欢迎畅所欲言！

**Comment by zepinglee** [comment]
> 采用新的代码风格和接口更多是出于可维护性的考虑：L3 引入的变量类型、展开类型符号以及库函数使得命令的执行过程更加清晰。

是的，像 `\thu@define@key` 其实是仿照 l3keys 的接口实现的。

> 我私下征求了当前维护者 @zepinglee 的意见，他认为应该新开分支逐步进行重构，以利用当前的 l3build 构建测试

理想的情况是，重构后的代码不改变 l3build 的 log-based test 输出（`/testfiles/**/*.tlg` 文件）。所以我建议分模块逐步重构，并且随时测试，可以方便定位 bug。如果将另起炉灶完全重构的代码导入进来再跑 l3build，可能有一大堆 diff，那就很难再定位问题了。

> 这样可以利用 DosStrip 的特性实现模块化的配置文件

我个人非常不喜欢 docstrip。

1. docstrip 的初衷是去掉注释以提高编译速度，但是在目前的计算机条件下并没有明显的性能优势。
2. dtx 有很多 `\begin{macrocode}` `\begin{macro}` 写起来很麻烦，读起来也很不方便。如果编译成 PDF 阅读，没有语法高亮更不方便。
3. 编译报错时无法定位到 .dtx 文件的行数，需要额外搜索代码。
4. 用户更新版本时需要覆盖 .dtx 文件，再编译 `.ins` 生成 `.cls`，但这很困扰不熟悉命令行的用户，目前相当一部分 issues 是用户没有生成新的 `.cls` 导致的。

**Reply by Harry-Chen** [reply]
我也不太喜欢 docstrip，如果要这些功能，我倒是觉得是否可以引入其他工具预处理一下（比如 Python），不一定拘泥于 LaTeX

**Reply by atxy-blip** [reply]
> 用户更新版本时需要覆盖 `.dtx` 文件

关于这点，我建议仓库里只储存用于开发的源代码，`thuthesis-example.tex` 等面向用户的文件随版本号更新打包上传到 Releases 区。

**Reply by zepinglee** [reply]
> 我也不太喜欢 docstrip，如果要这些功能，我倒是觉得是否可以引入其他工具预处理一下（比如 Python），不一定拘泥于 LaTeX

感觉可以。我觉得甚至可以按模块分开到多个文件中（比如封面、参考文献），在预处理时合并为一个 .cls 文件。

**Reply by zepinglee** [reply]
总之就是建议及时跑 l3build。我之前尝试用 latex3 改造 ustcthesis，很容易测试 fail，需要及时定位修正。

**Reply by atxy-blip** [reply]
> 是否可以引入其他工具预处理一下（比如 Python）

主观上不是很想造轮子（逃

L3 的名空间（@@ 模块标识）也是 DocStrip 提供的功能。`l3build`, `l3doc` 是一套完整的解决方案。

> 我建议仓库里只储存用于开发的源代码，`thuthesis-example.tex` 等面向用户的文件随版本号更新打包上传到 Releases 区。

补充一下，让用户 git clone 本仓库后直接用于论文写作会引发一些困惑（反复合并模板进度和写作进度）。我想的是做一个范式转变，提供一些一键式脚本，把 `.cls` 隐藏起来

- 让本地安装 TeX Live 的同学直接用 `tlmgr` 从 CTAN 拉取最新的的稳定版；
- 让克隆仓库的同学用 `l3build install` 安装到 TEXMF。这样就不用普及解包操作了。

在线的 Overleaf 和 TeXPage 使用前述的压缩包，目录下也能简洁一点。

**Reply by Harry-Chen** [reply]
> > 我建议仓库里只储存用于开发的源代码，`thuthesis-example.tex` 等面向用户的文件随版本号更新打包上传到 Releases 区。
> 
> 补充一下，让用户 git clone 本仓库后直接用于论文写作会引发一些困惑（反复合并模板进度和写作进度）。我想的是做一个范式转变，提供一些一键式脚本，把 `.cls` 隐藏起来
> 
> * 让本地安装 TeX Live 的同学直接用 `tlmgr` 从 CTAN 拉取最新的的稳定版；
> * 让克隆仓库的同学用 `l3build install` 安装到 TEXMF。这样就不用普及解包操作了。
> 
> 在线的 Overleaf 和 TeXPage 使用前述的压缩包，目录下也能简洁一点。

后面这个我赞同，看到很多人直接在 thuthesis 的基础上 commit，比较高血压

**Reply by zepinglee** [reply]
> 补充一下，让用户 git clone 本仓库后直接用于论文写作会引发一些困惑（反复合并模板进度和写作进度）。我想的是做一个范式转变，提供一些一键式脚本，把 `.cls` 隐藏起来
> 
> * 让本地安装 TeX Live 的同学直接用 `tlmgr` 从 CTAN 拉取最新的的稳定版；
> * 让克隆仓库的同学用 l3build install 安装到 TEXMF。这样就不用普及解包操作了。

我不太赞成。

`tlmgr` 不能跨年度版本更新宏包，比如 TeX Live 2022 不能更新 thuthesis 2023-04-07 v7.3.2。如果让用户每年都更新一次 TeX Live 也很不现实，尤其一些老师。

其次不同用户可能在 TEXMFHOME 安装了不同版本的 thuthesis，导致一份文档编译结果不同，会造成一些沟通上的成本。

总之最好 `.cls` 文件跟主文档在一起。

**Reply by note286** [reply]
我同意 @zepinglee 的说法。我觉得除了必要的logo等文件外，仅提供一个`.cls`文件，这对于用户来说是直观的、易于理解的。用户只需要编写`.tex`即可。当然，`.pdf`手册也是必备的。

手动安装至TeX Live对绝大部分用户是困惑的。

**Reply by xueruini** [reply]
至于用的时候怎么简单点，的确是个现实的问题。目前为了兼容各种需求，辅助文件比较多，但很多同学都不用tlmgr update，都是拷贝师兄师姐的目录过来。两难有点

**Comment by zepinglee** [comment]
@xueruini 意见如何？我记得在 <https://github.com/tuna/thuthesis/issues/424#issuecomment-511178441> 说过不太赞成 latex3 的方式。

**Reply by xueruini** [reply]
此一时彼一时。py3刚推的时候，难度也很大，大家觉得py2.7挺好。现在 l3 大势所趋，与时俱进是必须的 :)

btw，我觉得 docstrip 挺好玩儿，反正用 emacs 没觉得特别难受。就像用 l3 一样，社区的惯例就是选择
